pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -199,6 +245,27 @@ public void acknowledgeMessage(List<Position> positions, AckType ackType, Map<St
             if (log.isDebugEnabled()) {
                 log.debug("[{}][{}] Individual acks on {}", topicName, subName, positions);
             }
+
+            synchronized (PersistentSubscription.this) {
+                positions.forEach(position -> {
+                    // If single ack try to ack message in pending_ack status, fail the ack.
+                    if (this.pendingAckMessages.contains(position)) {
+                        log.warn("[{}][{}] Invalid acks position conflict with an ongoing transaction:{}{}.",
+                                topicName, subName, this.txnID.getMostSigBits(), this.txnID.getLeastSigBits());
+                        return;
+                    }
+
+                    // If single ack is within range of cumulative ack of an ongoing transaction, fail the ack.
+                    if (null != this.pendingCumulativeAckMessage &&
+                            ((PositionImpl) position).compareTo((PositionImpl) this.pendingCumulativeAckMessage) < 0) {
```

Comment:
do we need to check if `position` is an instance of `PositionImpl`?

Alternatively, maybe we can consider making `Position` extending `Comparable` interface?

Reply:
I thought we mostly check type in catch clause or passed parameter is an Object, and PositionImpl is only implementation class of Position so I skip the check.
But seems I do find some places in BK code base that we're checking if it's a PositionImpl instance. Also added here, just for safe.